### PR TITLE
Feat/2664 add argillatraining module for autotrain

### LINKED ISSUE #2664: add `argilla.training` module for `AutoTrain`
**Is your feature request related to a problem? Please describe.**
I want to fine-tune an `AutoTrain` model

**Describe the solution you'd like**
Fine-tune options for each corresponding task.

[] ENTITY EXTRACTION
[] CLASSIFICATION
[] TEXT2TEXT

**Describe alternatives you've considered**
N.A.

**Additional context**
https://github.com/argilla-io/argilla-streamlit/blob/main/my_app/pages/autotrain-project.py

### PR DESCRIPTION
**Pull request by davidberenstein1957**
# Description

added support for AutoTrain via direct Argilla dataset integrations.

Closes #2664

**Type of change**

(Please delete options that are not relevant. Remember to title the PR according to the type of change)

- [X] New feature (non-breaking change which adds functionality)
- [X] Refactor (change restructuring the codebase without changing functionality)
- [X] Improvement (change adding some improvement to an existing functionality)

**How Has This Been Tested**

pytest tests/training/test_autotrain.py

**Checklist**

- [ ] I added relevant documentation
- [X] follows the style guidelines of this project
- [X] I did a self-review of my code
- [X] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added relevant notes to the CHANGELOG.md file (See https://keepachangelog.com/)

### CODE REVIEW

File: src/argilla/training/autotrain_advanced.py [hunk 1]
```diff
@@ -0,0 +1,260 @@
...
+from uuid import uuid4
+
+from datasets import DatasetDict
+
+import argilla as rg
+from argilla.training.base import ArgillaTrainerSkeleton
+from argilla.utils.dependency import require_version
+
+
+class AutoTrainer(object):
```
Comment (alvarobartt) [inline src/argilla/training/autotrain_advanced.py]:
```suggestion
class AutoTrainMixIn:
```

File: src/argilla/training/autotrain_advanced.py [hunk 2]
```diff
@@ -0,0 +1,260 @@
...
+    def prepare_dataset(self, data_dict: dict = {}):
```
Comment (alvarobartt) [inline src/argilla/training/autotrain_advanced.py]:
```suggestion
    def prepare_dataset(self, data_dict_params: Optional[dict] = {}) -> None:
```

File: src/argilla/training/autotrain_advanced.py [hunk 3]
```diff
@@ -0,0 +1,260 @@
...
+        """
+        This function prepares a dataset for autotrain using a dictionary of data and specific column
+        mappings.
+
+        Args:
+          data_dict (dict): `data_dict` is a dictionary that contains additional parameters to be passed
+        to the `AutoTrainDataset` constructor. These parameters are optional and can be used to
+        customize the dataset preparation process.
```
Comment (alvarobartt) [inline src/argilla/training/autotrain_advanced.py]:
```suggestion
            to the `AutoTrainDataset` constructor. These parameters are optional and can be used to
            customize the dataset preparation process.
```

File: src/argilla/training/autotrain_advanced.py [hunk 4]
```diff
@@ -0,0 +1,260 @@
...
+
+
+class AutoTrainer(object):
+    def prepare_dataset(self, data_dict: dict = {}):
+        """
+        This function prepares a dataset for autotrain using a dictionary of data and specific column
+        mappings.
+
+        Args:
+          data_dict (dict): `data_dict` is a dictionary that contains additional parameters to be passed
```
Comment (alvarobartt) [inline src/argilla/training/autotrain_advanced.py]:
```suggestion
          data_dict_params: A dictionary that contains additional parameters to be passed
```

File: src/argilla/training/autotrain_advanced.py [hunk 5]
```diff
@@ -0,0 +1,260 @@
...
+        customize the dataset preparation process.
+        """
+        from autotrain.dataset import AutoTrainDataset
+
+        self.dset = AutoTrainDataset(
+            task=self.task,
+            token=self.HF_TOKEN,
+            username=self.AUTOTRAIN_USERNAME,
+            project_name=self.project_name,
+            **data_dict,
```
Comment (alvarobartt) [inline src/argilla/training/autotrain_advanced.py]:
```suggestion
            **data_dict_params,
```

File: src/argilla/training/autotrain_advanced.py [hunk 6]
```diff
@@ -0,0 +1,260 @@
...
+    def get_job_params(self):
+        if self._model.lower() == "autotrain":
+            job_params = self.trainer_kwargs["autotrain"]
+            model_choice = "autotrain"
+        else:
+            job_params = self.trainer_kwargs["hub_model"]
+            model_choice = "HuggingFace Hub"
+        if model_choice == "autotrain":
+            if len(job_params) > 1:
+                raise ValueError("❌ Only one job parameter is allowed for AutoTrain.")
```
Comment (alvarobartt) [inline src/argilla/training/autotrain_advanced.py]: Maybe we're more safe if we don't use emojis in the exceptions, so I'd reformat this as:

```suggestion
                raise ValueError("Only one job parameter is allowed for AutoTrain.")
```

File: src/argilla/training/autotrain_advanced.py [hunk 7]
```diff
@@ -0,0 +1,260 @@
...
+    require_version("autotrain-advanced")
+    require_version("datasets")
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+
+        self.project_name = f"{self._workspace}_{self._name}_{str(uuid4())[:8]}"
+
+        if self._seed:
+            self._logger.warning("❌ Setting a seed is not supported by `autotrain-advanced`.")
```
Comment (alvarobartt) [inline src/argilla/training/autotrain_advanced.py]: Same here with the emoji:

```suggestion
            self._logger.warning("Setting a seed is not supported by `autotrain-advanced`.")
```

File: src/argilla/training/autotrain_advanced.py [hunk 8]
```diff
@@ -0,0 +1,260 @@
...
+
+        Args:
+          text (Union[List[str], str]): The text to be classified.
+          as_argilla_records (bool): If True, the prediction will be returned as an Argilla record. If
+        False, the prediction will be returned as a string. Defaults to True
+
+        Returns:
+          A list of predictions
+        """
+        self._logger.error("Use framework=`transformers` for inference.")
```
Comment (alvarobartt) [inline src/argilla/training/autotrain_advanced.py]: As we have the information to do so, I'd show here an example on how to load the trained model/s using `transformers`, just to serve as a tip

File: src/argilla/training/autotrain_advanced.py [hunk 9]
```diff
@@ -0,0 +1,260 @@
...
+
+    def save(self, output_dir: str):
+        """
+        The function saves the model to the path specified, and also saves the label2id and id2label
+        dictionaries to the same path
+
+        Args:
+          output_dir (str): the path to save the model to
+        """
+        self._logger.warning("Models are saved on the Hugging Face hub, so this function is not supported.")
```
Comment (alvarobartt) [inline src/argilla/training/autotrain_advanced.py]:
```suggestion
        self._logger.warning("Models are saved on the HuggingFace Hub, so this function is not supported.")
```

File: src/argilla/training/base.py [hunk 10]
```diff
@@ -73,6 +73,9 @@ def __init__(
             **load_kwargs: arguments for the rg.load() function.
         """
         self._name = name
+        if workspace is None:
+            workspace = rg.get_workspace()
+        self._workspace = workspace
```
Comment (alvarobartt) [inline src/argilla/training/base.py]:
```suggestion
        self._workspace = workspace or rg.get_workspace()
```

File: src/argilla/training/autotrain_advanced.py [hunk 11]
```diff
@@ -0,0 +1,260 @@
...
+            if len(job_params) > 1:
+                raise ValueError("❌ Only one job parameter is allowed for AutoTrain.")
+            job_params[0].update({"task": self.task})
+        elif model_choice == "HuggingFace Hub":
+            for i in range(len(job_params)):
+                job_params[i].update({"task": self.task})
+        return job_params
+
+
+class ArgillaAutoTrainTrainer(ArgillaTrainerSkeleton, AutoTrainer):
```
Comment (alvarobartt) [inline src/argilla/training/autotrain_advanced.py]:
```suggestion
class ArgillaAutoTrainTrainer(ArgillaTrainerSkeleton, AutoTrainMixIn):
```

**Review by alvarobartt — COMMENTED** [review]
Some minor comments, but great work overall! 👍🏻 Ideally, those suggestions should be applied in batch, but check those carefully as I haven't tested those myself

File: src/argilla/training/base.py [hunk 12]
```diff
@@ -261,6 +292,8 @@ def save(self, output_dir: str):
 class ArgillaTrainerSkeleton(ABC):
     def __init__(
         self,
+        name: str,
+        workspace: str,
```
Comment (alvarobartt) [inline src/argilla/training/base.py]: I guess we can make this optional and default to `rg.get_workspace()`?

```suggestion
        workspace: Optional[str] = rg.get_workspace(),
```

Or the following instead:

```suggestion
        workspace: Optional[str] = None,
```

And then

```python
if workspace is None:
    workspace = rg.get_workspace()
...
```

WDYT? @davidberenstein1957

File: tests/training/test_autotrain.py [hunk 13]
```diff
@@ -0,0 +1,77 @@
...
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+import os
+
+import pytest
+from argilla.training import ArgillaTrainer
+
+FRAMEWORK = "autotrain"
+MODELS = ["prajjwal1/bert-tiny", "autotrain"]
+_HF_HUB_ACCESS_TOKEN = os.environ.get("HF_AUTH_TOKEN") or os.environ.get("HF_HUB_ACCESS_TOKEN")
```
Comment (alvarobartt) [inline tests/training/test_autotrain.py:21]: Here indeed we could mock it as I did for the `push_to_huggingface` and `from_huggingface` methods in `FeedbackDataset`, but we can have a look at this later together 😄

**Review by alvarobartt — COMMENTED** [review]
LGTM!

**Review by alvarobartt — COMMENTED** [review]
LGTM!

File: tests/training/test_autotrain.py (hunk 13 above)
Comment (davidberenstein1957) [reply tests/training/test_autotrain.py:21]: I agree!

**Comment by davidberenstein1957** [comment]
@frascuchon do you want me to merge this in dev or release?

**Pull request merged**